### PR TITLE
Implements CoinGecko DataSource

### PR DESCRIPTION
**Pull request by AlexCatarino**

### CODE REVIEW

File: QuantConnect.DataSource.csproj [hunk 1]
```diff
@@ -2,14 +2,14 @@
...
     <RootNamespace>QuantConnect.DataSource</RootNamespace>
-    <AssemblyName>QuantConnect.DataSource.MyCustomDataType</AssemblyName>
+    <AssemblyName>QuantConnect.DataSource.CoinGecko</AssemblyName>
     <OutputPath>bin\$(Configuration)</OutputPath>
-    <DocumentationFile>$(OutputPath)\QuantConnect.DataSource.MyCustomDataType.xml</DocumentationFile>
+    <DocumentationFile>$(OutputPath)\QuantConnect.DataSource.CoinGecko.xml</DocumentationFile>
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="QuantConnect.Algorithm.Framework" Version="2.5.*" />
```
Comment (Martin-Molinero) [inline QuantConnect.DataSource.csproj:11]: `QuantConnect.Algorithm.Framework` isn't required I think?

File: DataProcessing/DataProcessing.csproj [hunk 2]
```diff
@@ -6,24 +6,9 @@
         <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
     </PropertyGroup>
     <ItemGroup>
-        <PackageReference Include="QuantConnect.Lean.Engine" Version="2.5.*"/>
-        <PackageReference Include="QuantConnect.Research" Version="2.5.*"/>
-        <PackageReference Include="QuantConnect.pythonnet" Version="2.0.*" />
-        <PackageReference Include="fasterflect" Version="3.0.0" />
+        <PackageReference Include="QuantConnect.Lean.Engine" Version="2.5.*" />
```
Comment (Martin-Molinero) [inline DataProcessing/DataProcessing.csproj]: We probably only need common nuget?

File: DataProcessing/Program.cs [hunk 3]
```diff
@@ -37,17 +37,18 @@ public static void Main()
             var destinationDirectory = Path.Combine(
                 Config.Get("temp-output-directory", "/temp-output-directory"),
                 "alternative",
-                "vendorname");
+                "coingecko");
+            var processedDirectory = Path.Combine(Globals.DataFolder, "alternative", "coingecko");
```
Comment (Martin-Molinero) [inline DataProcessing/Program.cs]: Believe can replace for `coingecko` `CoinGeckoUniverseDataDownloader.VendorName`

File: CoinGecko.cs [hunk 4]
```diff
@@ -16,34 +16,43 @@
...
 using System.IO;
 using QuantConnect.Data;
 using System.Collections.Generic;
+using System.Globalization;
 
 namespace QuantConnect.DataSource
 {
     /// <summary>
-    /// Example custom data type
+    /// Universe Selection for Coin Gecko data which contains Price, Volume and Market Cap
```
Comment (Martin-Molinero) [inline CoinGecko.cs]: > contains Price

Price of the coin in USD or against what currency? think we should clarify it 👍 same for the universe version

> Universe Selection

This comment seems just for the universe version?

File: CoinGecko.cs [hunk 5]
```diff
@@ -92,65 +102,60 @@ public override BaseData Reader(SubscriptionDataConfig config, string line, Date
...
         /// </summary>
-        /// <returns>false</returns>
-        public override bool RequiresMapping()
+        /// <param name="market">The market the ticker resides in</param>
+        /// <param name="quoteCurrency">The quote currency of the crypto-currency pair. E.g. USD for BTCUSD</param>
+        /// <returns>A new Symbol object for the specified ticker</returns>
+        public Symbol CreateSymbol(string market, string quoteCurrency = "USD")
         {
-            return true;
+            return Symbol.Create($"{Coin}{quoteCurrency}", SecurityType.Crypto, market);
```
Comment (Martin-Molinero) [inline CoinGecko.cs]: Think `SecurityType` should probably be an optional argument too, thinking about the crypto futures potential use case here

File: DataProcessing/CoinGeckoDataDownloader.cs [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+        /// Control the rate of download per unit of time.
+        /// </summary>
+        private readonly RateGate _indexGate;
+
+        /// <summary>
+        /// Creates a new instance of <see cref="CoinGeckoUniverseDataDownloader"/>
+        /// </summary>
+        /// <param name="destinationFolder">The folder where the data will be saved</param>
+        /// <param name="apiKey">The Vendor API key</param>
+        public CoinGeckoUniverseDataDownloader(string destinationFolder, string processedFolder, string apiKey = null)
```
Comment (Martin-Molinero) [inline DataProcessing/CoinGeckoDataDownloader.cs]: >  string apiKey = null

Isn't used, should we just remove it then?
Missing xml docs for `processedFolder`

File: DataProcessing/config.json [hunk 7]
```diff
@@ -1,5 +0,0 @@
-{
-    "data-folder": "../../../Data/",
```
Comment (Martin-Molinero) [inline DataProcessing/config.json:2]: > "data-folder"

Is a config required by the data processor, suggest leaving the config.json with the needed configurations

File: DataProcessing/CoinGeckoDataDownloader.cs [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+        {
+            _destinationFolder = destinationFolder;
+            _processedFolder = processedFolder;
+            _universeFolder = Path.Combine(_destinationFolder, "universe");
+            _processedUniverseFolder = Path.Combine(_processedFolder, "universe");
+
+            // CoinGecko: Our Free API* has a rate limit of 10-50 calls/minute
+            // Represents rate limits of 5 requests per 1 minute
```
Comment (Martin-Molinero) [inline DataProcessing/CoinGeckoDataDownloader.cs]: > 10-50 calls/minute

Seems quite higher than 5/minute we are using?

File: DataProcessing/CoinGeckoDataDownloader.cs [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+        private CoinGeckoMarketChart GetCoinGeckoMarketChartsForId(string id, bool exists)
+        {
+            string value;
+            if (File.Exists($"{id}.json"))
+            {
+                value = File.ReadAllText($"{id}.json");
+            }
+            else
+            {
+                var days = exists ? "1" : "max";
```
Comment (Martin-Molinero) [inline DataProcessing/CoinGeckoDataDownloader.cs]: Is `max` enough to cover the whole history? for example for BTC, would be nice to add a comment of what max is

**Review by Martin-Molinero — COMMENTED** [review]
Cool! posted a couple of comments 👍

File: DataProcessing/CoinGeckoDataDownloader.cs (hunk 9 above)
Comment (Martin-Molinero) [reply DataProcessing/CoinGeckoDataDownloader.cs]: Seems it is, BTC data is complete 👍

File: QuantConnect.DataSource.csproj (hunk 1 above)
Comment (LouisSzeto) [reply QuantConnect.DataSource.csproj:11]: We got `using QuantConnect.Algorithm.Framework.Selection;` in `CoinGeckoUniverseSelectionModel.cs` that requires it

**Pull request merged**